### PR TITLE
cache based proving

### PR DESCRIPTION
**Pull request by noamnelke**

### CODE REVIEW

**Comment by noamnelke** [comment]
The hashes (roots, proofs) and benchmarks have changed because I changed the leaf size from 8 to 32 bytes.

File: proving.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+				// Prepare the reader for traversing the subtree.
+				err := reader.Seek(subtreeStart.index)
+				if err != nil {
+					return nil, errors.New("while seeking in cache: " + err.Error() + subtreeStart.String())
+				}
+
+				// Traverse the subtree.
+				width = 1 << (currentPos.height - subtreeStart.height)
+				_, root, err := traverseSubtree(reader, width, hash, nil)
```
Comment (zalmen) [inline proving.go]: `root` is the value of `currentPos`, right? If true, it might help to mention it in a comment

File: proving.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+				break
+			}
+
+			// Add the current node sibling to the proof:
+
+			// Get the cache reader for the current layer.
+			reader, found = readers[currentPos.height]
+
+			// If the cache wan't found, we calculate the minimal subtree that will get us the required node.
+			if !found {
```
Comment (zalmen) [inline proving.go]: All of the if clause should be extracted to a separate method, preferably it should be a method of a cache struct

File: proving.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+type NodeReader interface {
+	Seek(index uint64) error
+	ReadNext() ([]byte, error)
+	Width() uint64
+}
+
+func GenerateProof(
+	provenLeafIndices []uint64,
+	readers map[uint]NodeReader,
```
Comment (zalmen) [inline proving.go]: There is quite a lot of logic around accessing and maintaining the cache which justifies building a dedicated cache struct with supporting methods

File: proving.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+		subtreeStart, currentPos, width = subtreeDefinition(currentPos, readers)
+
+		// Prepare reader to read subtree leaves.
+		err = reader.Seek(subtreeStart.index)
+		if err != nil {
+			return nil, errors.New("while preparing to traverse subtree: " + err.Error())
+		}
+
+		// Prepare list of leaves to prove in the subtree.
+		leavesToProve := provenLeafIndexIt.batchPopRelativeIndices(subtreeStart.index, width)
```
[the remark this answers is not in the document]
Comment (noamnelke) [reply proving.go]: Will extract the code that makes it "relative" to outside this method.

File: proving.go [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+// subtreeDefinition returns the definition (firstLeaf and root positions, width) for the minimal subtree whose
+// base layer includes p and where the root is on a cached layer.
+func subtreeDefinition(p position, readers map[uint]NodeReader) (firstLeaf, root position, width uint64) {
+	// maxRootHeight represents the max height of the tree, based on the width of base layer. This is used to prevent an
+	// infinite loop.
+	maxRootHeight := rootHeightFromWidth(readers[p.height].Width())
+	for root = p.parent(); root.height < maxRootHeight; root = root.parent() {
+		if _, found := readers[root.height]; found {
```
[the remark this answers is not in the document]
Comment (noamnelke) [reply proving.go]: Not sure if this is still relevant after the refactor. I don't think we discussed it this morning - if it's still relevant, please elaborate (or we can talk tomorrow).

File: proving.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+	provenLeafIndexIt := &positionsIterator{s: provenLeafIndices}
+	skipPositions := &positionsStack{}
+	rootHeight := rootHeightFromWidth(readers[0].Width())
+
+	cache, err := NewTreeCache(readers, hash)
+	if err != nil {
+		return nil, err
+	}
+
+	for ; ; currentPos = currentPos.parent() { // Process proven leaves:
```
Comment (zalmen) [inline proving.go]: don't understand why you need to "increment" `currentPos`.. Anyway you assign it with a new value when calling `subtreeDefinition`

File: proving.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zalmen) [reply proving.go]: It's no longer relevant

File: proving.go [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+	}
+
+	return &TreeCache{
+		readers: readers,
+		hash:    hash,
+	}, nil
+}
+
+// subtreeDefinition returns the definition (firstLeaf and root positions, width) for the minimal subtree whose
+// base layer includes p and where the root is on a cached layer.
```
Comment (zalmen) [inline proving.go]: correction: "the root is either on a cached layer or the original tree's root"

File: proving.go [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+	var currentPos, subtreeStart position
+	var width uint64
+	var proof, additionalProof [][]byte
+
+	provenLeafIndexIt := &positionsIterator{s: provenLeafIndices}
+	skipPositions := &positionsStack{}
+	rootHeight := rootHeightFromWidth(readers[0].Width())
+
+	cache, err := NewTreeCache(readers, hash)
```
Comment (zalmen) [inline proving.go]: Why do you still need to hold the readers map? Why not have the `TreeCache` as the input type of the cache without using the map directly (you can add a method to `TreeCache` that returns a `NodeReader` for a given height)

File: merkle.go [hunk 9]
```diff
@@ -33,12 +34,19 @@ func newLayer(height uint, cache io.Writer) *layer {
 	return &layer{height: height, cache: cache}
 }
 
-type sparseBoolStack struct {
+type SparseBoolStack struct {
```
[the remark this answers is not in the document]
Comment (noamnelke) [reply merkle.go]: No, but exporting it was a mistake. I'm making it lowercase again and not adding tests.

File: proving.go (hunk 6 above)
Comment (noamnelke) [reply proving.go]: Good point, it's only needed in the inner loop. Fixing.

File: proving.go (hunk 8 above)
Comment (noamnelke) [reply proving.go]: I don't really "hold" it, it's just in the local scope. I did *use* it, however, which is essentially the same. I now extracted this logic into the cache, as well (`calcProofForNextLeaf`).

File: proving.go [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+	// Traverse the subtree.
+	width := uint64(1) << (nodePos.height - subtreeStart.height)
+	_, currentVal, err := traverseSubtree(reader, width, c.hash, nil)
+	if err != nil {
+		return nil, errors.New("while traversing subtree for root: " + err.Error())
+	}
+	return currentVal, nil
+}
+
+func (c *TreeCache) calcProofForNextLeaf(nextProvenLeafPos position, provenLeafIndexIt *positionsIterator) ([][]byte,
```
Comment (zalmen) [inline proving.go]: this method shouldn't be under `TreeCache`. It uses the cache but it's far from serving a cache logic.. Make this method a function and rename it to `calcProofForLeaf` (the "next" is a context known only to the caller, the function doesn't care about that fact).

**Pull request merged**